persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala

```diff
@@ -0,0 +1,128 @@
+import com.typesafe.config.Config
+
+import scala.concurrent.duration._
+
+/**
+ * Settings for cluster distribution.
+ *
+ * @param clusterShardingSettings The cluster sharding settings.
+ * @param ensureActiveInterval The interval at which entities are ensured to be active.
+ */
+case class ClusterDistributionSettings(
+  clusterShardingSettings: ClusterShardingSettings,
+  ensureActiveInterval:    FiniteDuration
+)
+
+object ClusterDistributionSettings {
+  def apply(system: ActorSystem): ClusterDistributionSettings = {
+    val clusterShardingSettings = ClusterShardingSettings(system)
+    val config = system.settings.config.getConfig("akka.cluster.distribution")
+    ClusterDistributionSettings(config, clusterShardingSettings)
+  }
+
+  def apply(config: Config, clusterShardingSettings: ClusterShardingSettings): ClusterDistributionSettings = {
+    val ensureActiveInterval = config.getDuration("ensure-active-interval", TimeUnit.MILLISECONDS).milliseconds
+    ClusterDistributionSettings(clusterShardingSettings, ensureActiveInterval)
+  }
+}
+
+object ClusterDistribution extends ExtensionId[ClusterDistribution] with ExtensionIdProvider {
+
+  override def createExtension(system: ExtendedActorSystem): ClusterDistribution =
+    new ClusterDistribution(system)
+
+  override def lookup = ClusterDistribution
+
+  /**
+   * Sent to each entity in a cluster distribution to ensure it's active.
+   */
+  case class EnsureActive(entityId: EntityId)
+}
+
+/**
+ * Distributes a static list of entities evenly over a cluster, ensuring that they are all continuously active.
+ *
+ * This uses cluster sharding underneath, using a scheduled tick sent to each entity as a mechanism to ensure they are
+ * active.
+ *
+ * Entities are cluster sharding entities, so they can discover their ID by inspecting their name. Additionally,
+ * entities should handle the [[com.lightbend.lagom.internal.persistence.cluster.ClusterDistribution.EnsureActive]]
+ * message, typically they can do nothing in response to it.
+ */
+class ClusterDistribution(system: ExtendedActorSystem) extends Extension {
+
+  /**
+   * Start a cluster distribution.
+   *
+   * @param typeName The name of the type of entity. This is used as the cluster sharding type name.
+   * @param entityProps The props for the entity actor.
+   * @param entityIds The entity ids to distribute over the cluster.
+   * @param settings The cluster distribution settings.
+   * @return the actor ref of the [[ShardRegion]] that is to be responsible for the shard
+   */
+  def start(
+    typeName:    String,
+    entityProps: Props,
+    entityIds:   Seq[EntityId],
+    settings:    ClusterDistributionSettings
+  ): ActorRef = {
+
+    val extractEntityId: ShardRegion.ExtractEntityId = {
+      case msg @ EnsureActive(entityId) => (entityId, msg)
+    }
+    val extractShardId: ShardRegion.ExtractShardId = {
+      case EnsureActive(entityId) => entityId
```

Comment:
> why isn't ExtractShardId a function of entity id -> shard id

Yeah, that would be enough for almost all cases. The idea was that it should be possible to derive the shard id from more properties than only the entity id.

I see no reason for custom messages here, use `Identify`.

It wouldn't hurt to cap the number of shards with something like `entityId % 1000`. Doesn't have to be configurable.